File: Application/src/main/java/by/gastrofest/service/ParserService.java
package by.gastrofest.service;

import by.gastrofest.dbo.GastroFestDbo;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.springframework.stereotype.Service;

import static by.gastrofest.constant.MainConstants.MAIN_PAGE_URL;
import static by.gastrofest.constant.MainConstants.NODE_RECORD_CLASS;
import static by.gastrofest.utils.HttpUtil.getDocument;

@Service
@RequiredArgsConstructor
public class ParserService {

    private final GastroSetService gastroSetService;

    private final ParticipantService participantService;

    private final GastrofestService gastrofestService;

    @Transactional
    public void parseMainPage() {
        final var document = getDocument(MAIN_PAGE_URL);
        final var savedGastrofest = saveGastrofestDbo(document);

        final var participantsNodes = document.getElementsByClass(NODE_RECORD_CLASS);
        for (final Element participantsNode : participantsNodes) {
            saveGastroSetDbo(participantsNode, savedGastrofest);
        }
    }

    private GastroFestDbo saveGastrofestDbo(final Document document) {
        final var gastrofestDbo = gastrofestService.extractGastrofestFromElement(document);
        return gastrofestService.save(gastrofestDbo);
    }

    private void saveGastroSetDbo(final Element participantsNode, final GastroFestDbo savedGastrofest) {
        final var gastroSetDbo = gastroSetService.extractGastroSetInfoFromMainPage(participantsNode);
        final var gastroSetDocument = getDocument(gastroSetDbo.getUrl());
        gastroSetService.updateGastroSetFromGastroSetPage(gastroSetDocument, gastroSetDbo);

        var participant = participantService.getParticipantFromGastroSetPage(gastroSetDocument);
        participant = participantService.save(participant);

        gastroSetDbo.setGastrofest(savedGastrofest);
        gastroSetDbo.setParticipant(participant);
        gastroSetService.save(gastroSetDbo);
    }
}

File: Parser/src/main/kotlin/by/gastrofest/parser/service/ParserService.kt
package by.gastrofest.parser.service

import by.gastrofest.parser.constant.NODE_RECORD_CLASS
import by.gastrofest.parser.constant.SCRAPE_URL
import by.gastrofest.parser.getDocument
import by.gastrofest.parser.model.GastroFest
import by.gastrofest.parser.model.GastroSet
import by.gastrofest.parser.model.Participant
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.select.Elements

class ParserService(
    private val gastrofestCommonService: GastrofestCommonService
) {

    fun getGastroSetsList3(): List<GastroSet> {
        val objectMapper = jacksonObjectMapper().apply {
            registerModule(JavaTimeModule())
            disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
        }
        val inputStream = Thread.currentThread()
            .contextClassLoader
            .getResourceAsStream("10-th.Gastrofest.json")
            ?: error("File not found in resources")

        val json = inputStream.bufferedReader().readText()
        return objectMapper.readValue(json)
    }

    fun getGastroSetsList(): List<GastroSet> {
        val document: Document = getDocument(SCRAPE_URL)
        val gastroFest: GastroFest = extractGastrofestFromElement(document)
        if (!gastrofestCommonService.shouldContinue(gastroFest)) {
            return emptyList()
        }

        val participantsNodes: Elements = document.getElementsByClass(NODE_RECORD_CLASS)
        return participantsNodes.map { parseGastroSet(it, gastroFest) }
    }

    private fun parseGastroSet(participantsNode: Element, savedGastrofest: GastroFest): GastroSet {
        val gastroSet: GastroSet = extractGastroSetInfoFromMainPage(participantsNode)
        val gastroSetDocument: Document = getDocument(gastroSet.url)
        updateGastroSetFromGastroSetPage(gastroSetDocument, gastroSet)

        val participant: Participant = getParticipantFromGastroSetPage(gastroSetDocument)

        gastroSet.gastrofest = savedGastrofest
        gastroSet.participant = participant
        return gastroSet
    }
}
